lib/Doctrine/ORM/PersistentCollection.php

```diff
@@ -692,23 +692,41 @@ public function unwrap()
     protected function doInitialize()
     {
         // Has NEW objects added through add(). Remember them.
-        $newObjects = [];
+        $newlyAddedDirtyObjects = [];
 
         if ($this->isDirty) {
-            $newObjects = $this->collection->toArray();
+            $newlyAddedDirtyObjects = $this->collection->toArray();
         }
 
         $this->collection->clear();
         $this->em->getUnitOfWork()->loadCollection($this);
         $this->takeSnapshot();
 
+        if ($newlyAddedDirtyObjects) {
+            $this->restoreNewObjectsInDirtyCollection($newlyAddedDirtyObjects);
+        }
+    }
+
+    /**
+     * @param object[] $newObjects
+     *
+     * @return void
+     *
+     * Note: the only reason why this entire looping/complexity is performed via `spl_object_hash`
+     *       is because we want to prevent using `array_udiff()`, which is likely to cause very
+     *       high overhead (complexity of O(n^2)). `array_diff_key()` performs the operation in
+     *       core, which is faster than using a callback for comparisons
+     */
+    private function restoreNewObjectsInDirtyCollection(array $newObjects)
+    {
+        $loadedObjects               = $this->collection->toArray();
+        $newObjectsByOid             = array_combine(array_map('spl_object_hash', $newObjects), $newObjects);
+        $loadedObjectsByOid          = array_combine(array_map('spl_object_hash', $loadedObjects), $loadedObjects);
+        $newObjectsThatWereNotLoaded = array_diff_key($newObjectsByOid, $loadedObjectsByOid);
+
         // Reattach NEW objects added through add(), if any.
-        if ($newObjects) {
-            foreach ($newObjects as $obj) {
-                $this->collection->add($obj);
-            }
+        array_walk($newObjectsThatWereNotLoaded, [$this->collection, 'add']);
 
-            $this->isDirty = true;
-        }
+        $this->isDirty = true;
```

Comment:
Assume you add the following code after https://github.com/doctrine/doctrine2/commit/1fc7f81741be84bbce6f92c61d6f6c2b4a3e573f#diff-c32ade803f1349ec4da7da8644879ef6R65:
```
        $this->_em->flush();

        self::assertFalse($phones->isInitialized());
        self::assertFalse($phones->isDirty());

        $phones->add($item2);

        self::assertFalse($phones->isInitialized());
        self::assertTrue($phones->isDirty());

        $phones->initialize();
 
        self::assertTrue($phones->isInitialized());
        self::assertFalse($phones->isDirty());
```

After the initialize, the only item that was added to the uninitialized collection was already present after the collection was initialized. Thus, after initializing, the collection is not dirty.